Allocate floating ips from the pool instead of relying on pre-allocated ones

Fodoj opened the pull request:
I am not entirely sure why current implementation gets free IP from the pool and hopes there are actually already pre-allocated IPs in this pool.

I feel it makes more sense to allocate a new IP dynamically and release it once the test is done. 

Otherwise one would need to rely on manually pre-allocated IPs which is not a proper automation, I believe.

Note that with this change floating_ip_pool must a network ID, not network name.

Fodoj commented:
Any update on this?

jjasghar commented:
Looking at it, it looks ok. I need to validate it, or find someone else to 👍  this before i'm willing to merge this.

Fodoj commented:
@jjasghar great, thanks. I rebased commits, fixed all rubocop warnings and dropped ruby 2.0 from Travis config. Travis installs fresh gems before the build and they require 2.1+. I tried to fix gem versions and it worked, till it appeared that the last ruby 2.0 compatible ffi gem is not compatible with ruby 2.3.1.. bummer. And ruby 2.0 is obsolete anyways, so I assumed it's safe to just drop it from the build.

cliles commented:
This would be a great addition!

Fodoj commented:
@jjasghar did you get a chance to test it?

cliles commented:
@Fodoj - I didn't make it clear in my +1, but I did test it. This works as advertised. The only thing I can add is that it would be nice to not introduce a change to floating_ip_pool. Maybe add a lookup to find the id from the name first?

Fodoj commented:
I am not sure which benefit would it bring except backwards compatibility

Fodoj commented:
@cliles so should I implemented it to get this PR merge or its good as it is?

cliles commented:
I am just some guy on the internet, not a team-member. It's my opinion that it should be as backwards compatible as possible, or use a feature flag if new behavior is introduced.

Either way, I'd love to see this merged and a new release.

Maybe @jjasghar would like to comment, or someone from team test-kitchen (@Atalanta @calebhailey @cmluciano @erjohnso @fnichol @gregf @jtimberman @mbrukman @mwrock @robbkidd @RoboticCheese @schisamo @sethvargo @tas50 @Temikus @tyler-ball @webframp).

jjasghar commented:
@Fodoj per @cliles I'd suggest building in the backwards compatibility. After that, I'm good with this and will merge/release.  Thanks for the addition!

quulah commented:
Related to this, if you run multiple tests for multiple cookbooks simultaneously the tests might fail.

I'm guessing that the driver now checks for unassociated floating IPs from the pool and depending on the timing, two test runs might get the same IP and fight over it. At least that's how it looks like in the logs.

Allocating a totally new IP, associating it with the to-be tested instance and then throwing it away when the tests are done will probably solve the problems we're having too.

dfairhurst commented:
Patiently waiting for this feature :) :+1:

jjasghar commented:
I'll revisit this asap.

dannytrigo commented:
I opened https://github.com/test-kitchen/kitchen-openstack/pull/155 to add a flag to turn on the new behaviour

jjasghar commented:
Please rebase off the upstream changes.

jjasghar commented:
Closing due to conflicts and i think #155 does it for this.
